### PR TITLE
fix(helmfile): support case with oci repository in different document

### PR DESCRIPTION
**Pull request by PhilipAbed**
## Changes

reading the `oci:true` flag in the `helmfile.yaml`

before: it was reading only in the same document
after: it will be read as part of the repositoryData, and then used later in the releases

## Context

https://github.com/renovatebot/renovate/discussions/30168

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/helmfile/extract.ts [hunk 1]
```diff
@@ -106,23 +110,21 @@ export async function extractPackageFile(
...
-      );
+
       if (isOCIRegistry(dep.chart)) {
         res.datasource = DockerDatasource.id;
         res.packageName = repoName + '/' + depName;
-      } else if (repository?.oci) {
+      } else if (registryData[repoName]?.oci) {
         res.datasource = DockerDatasource.id;
-        res.packageName = registryAliases[repoName] + '/' + depName;
+        res.packageName = registryData[repoName]?.url
```
Comment (viceice) [inline lib/modules/manager/helmfile/extract.ts]: use string template or urlJoin helper

File: lib/modules/manager/helmfile/extract.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ import type {
   PackageFileContent,
 } from '../types';
 import type { Doc } from './schema';
-import { Doc as documentSchema } from './schema';
+import {Doc as documentSchema, HelmRepository} from './schema';
```
Comment (viceice) [inline lib/modules/manager/helmfile/extract.ts]:
```suggestion
import {Doc as documentSchema, type HelmRepository} from './schema';
```

File: lib/modules/manager/helmfile/extract.ts [hunk 3]
```diff
@@ -51,11 +51,15 @@ export async function extractPackageFile(
...
     return null;
   }
+
   for (const doc of docs) {
     // Always check for repositories in the current document and override the existing ones if any (as YAML does)
     if (doc.repositories) {
-      registryAliases = {};
+      registryData = {};
+      // used for printing only
+      let registryAliases: Record<string, string> = {};
```
Comment (viceice) [inline lib/modules/manager/helmfile/extract.ts]: why not printing the new object to logs?

File: lib/modules/manager/helmfile/extract.ts [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import is from '@sindresorhus/is';
+import urlJoin from 'url-join';
```
Comment (secustor) [inline lib/modules/manager/helmfile/extract.ts:2]: use our own abstraction
```suggestion
import { joinUrlParts } from '../../../util/url';
```

File: lib/modules/manager/helmfile/extract.ts [hunk 5]
```diff
@@ -51,15 +52,16 @@ export async function extractPackageFile(
     );
     return null;
   }
+
   for (const doc of docs) {
     // Always check for repositories in the current document and override the existing ones if any (as YAML does)
     if (doc.repositories) {
-      registryAliases = {};
+      registryData = {};
```
Comment (secustor) [inline lib/modules/manager/helmfile/extract.ts:59]:
```suggestion
```
You are resetting `registryData` iteration, that way you will have only the data for that specific document again. 🤔

Comment (viceice) [reply lib/modules/manager/helmfile/extract.ts:59]: expected if the current document is defining it's own repos

Comment (secustor) [reply lib/modules/manager/helmfile/extract.ts:59]: But can't a document define repos and reference ones in a other document at the same time?

Comment (viceice) [reply lib/modules/manager/helmfile/extract.ts:59]: I don't think so. if I remember correctly it's always overriding the previous.

Comment (PhilipAbed) [reply lib/modules/manager/helmfile/extract.ts:59]:
```
or (const doc of docs) {
    // Always check for repositories in the current document and override the existing ones if any (as YAML does)
```
it says in the comment

File: lib/modules/manager/helmfile/extract.spec.ts [hunk 6]
```diff
@@ -363,6 +363,36 @@ describe('modules/manager/helmfile/extract', () => {
...
+      const content =
+        'repositories:\n' +
+        '  - name: oci-repo\n' +
+        '    url: ghcr.io/example/oci-repo\n' +
+        '    oci: true\n' +
+        '---\n' +
+        'releases:\n' +
+        '  - name: example\n' +
+        '    version: 0.1.0\n' +
+        '    chart: oci-repo/example';
```
Comment (secustor) [inline lib/modules/manager/helmfile/extract.spec.ts:376]: use a codeBlock instead
https://github.com/renovatebot/renovate/blob/e2a7586466df0370ac71659332b5fb459a483d4b/lib/util/toml.spec.ts#L6-L10

**Review by secustor — COMMENTED** [review]
small nitpick, else LGTM

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/helmfile/extract.spec.ts (hunk 6 above)
Comment (PhilipAbed) [reply lib/modules/manager/helmfile/extract.spec.ts:376]: eh, last time i tried that, it failed on indentations :/ ill try

File: lib/modules/manager/helmfile/extract.spec.ts [hunk 7]
```diff
@@ -363,6 +363,37 @@ describe('modules/manager/helmfile/extract', () => {
...
+repositories:
+  - name: oci-repo
+    url: ghcr.io/example/oci-repo
+    oci: true
+---
+releases:
+  - name: example
+    version: 0.1.0
+    chart: oci-repo/example
+`;
```
Comment (secustor) [inline lib/modules/manager/helmfile/extract.spec.ts:377]:
```suggestion
      const content = codeBlock`
      repositories:
        - name: oci-repo
          url: ghcr.io/example/oci-repo
          oci: true
      ---
      releases:
        - name: example
          version: 0.1.0
          chart: oci-repo/example
      `;
```

**Review by secustor — APPROVED** [review]

**Pull request merged**